### PR TITLE
Fix AWS ECR private repository usage

### PR DESCRIPTION
**Pull request by ErikEvenson**
This is my attempt to allow nomad to access AWS ECR private docker repositories.  The problem is per the commit discussion between me and @diptanu here: https://github.com/hashicorp/nomad/commit/c4fd2360261fd6199af97e0fbcd83d2eb0ba6a67

This solution does require the user to add `https://` to the image name in the docker config.  I am not sure if that is in line with nomad's design vision.  Other declarative approaches are possible, but this does work for my AWS ECR usage.

### CODE REVIEW

**Comment by ErikEvenson** [comment]
Here is a job that uses this solution successfully:

```
job "vbez-web" {
	region = "global"
	datacenters = ["dc0"]
	type = "service"
	priority = 50

	constraint {
		attribute = "${attr.kernel.name}"
		value = "linux"
	}

	update {
		stagger = "10s"
		max_parallel = 1
	}

	group "vbez-web" {
		count = 1

		restart {
			interval = "5m"
			attempts = 10
			delay = "25s"
			mode = "delay"
		}

		task "api" {
			driver = "docker"

			config {
				dns_servers = ["172.17.0.1", "75.75.76.76", "75.75.75.75"]
				image = "https://212181198242.dkr.ecr.us-east-1.amazonaws.com/vbenergy/vbez-web:2016.02.19.0"
				port_map {
					web = 8080
				}
			}

			logs {
			  max_files = 10
			  max_file_size = 10
			}

			service {
				name = "vbez-web"
				tags = ["global"]
				port = "web"

				check {
					name = "alive"
					type = "tcp"
					interval = "10s"
					timeout = "2s"
				}
			}

			resources {
				cpu = 500
				disk = 101
				memory = 256
				network {
					mbits = 10
					port "web" {
					}
				}
			}
		}
	}
}

```

**Comment by ErikEvenson** [comment]
I am not sure I understand the CI failure here.  The client driver tests pass on my local machine.

**Comment by dadgar** [comment]
@ErikEvenson That test has become flakey recently, don't worry about that one.

**Comment by ErikEvenson** [comment]
Great -- I'll take that as good news.

File: client/driver/docker.go [hunk 1]
```diff
@@ -408,6 +409,14 @@ func (d *DockerDriver) Start(ctx *ExecContext, task *structs.Task) (DriverHandle
 	if err := mapstructure.WeakDecode(task.Config, &driverConfig); err != nil {
 		return nil, err
 	}
+
+	if strings.Contains(driverConfig.ImageName, "https://") {
+		driverConfig.SSL = true
+		driverConfig.ImageName = strings.Replace(driverConfig.ImageName, "https://", "", 1)
+	} else {
+		driverConfig.SSL = false
```
Comment (diptanu) [inline client/driver/docker.go]: I think we don't need to set this as false as default value for bools are ```false```

File: client/driver/docker.go [hunk 2]
```diff
@@ -408,6 +409,14 @@ func (d *DockerDriver) Start(ctx *ExecContext, task *structs.Task) (DriverHandle
 	if err := mapstructure.WeakDecode(task.Config, &driverConfig); err != nil {
 		return nil, err
 	}
+
+	if strings.Contains(driverConfig.ImageName, "https://") {
```
Comment (diptanu) [inline client/driver/docker.go]: Can we move this logic out of ```validate``` into the createContainer method?

File: client/driver/docker.go [hunk 3]
```diff
@@ -60,6 +60,7 @@ type DockerDriverConfig struct {
 	LabelsRaw        []map[string]string `mapstructure:"labels"`             //
 	Labels           map[string]string   `mapstructure:"-"`                  // Labels to set when the container starts up
 	Auth             []DockerDriverAuth  `mapstructure:"auth"`               // Authentication credentials for a private Docker registry
+	SSL              bool                `mapstructure:"ssl"`
```
Comment (diptanu) [inline client/driver/docker.go]: Please add a comment here.

File: client/driver/docker.go [hunk 4]
```diff
@@ -471,7 +480,14 @@ func (d *DockerDriver) Start(ctx *ExecContext, task *structs.Task) (DriverHandle
...
 					return nil, fmt.Errorf("Failed to create docker auth object: %v", err)
 				}
-				if authConfiguration, ok := authConfigurations.Configs[repo]; ok {
+
+				authConfigurationKey := ""
+				if driverConfig.SSL {
+					authConfigurationKey += "https://"
+				}
+
+				authConfigurationKey += strings.Split(driverConfig.ImageName, "/")[0]
```
Comment (diptanu) [inline client/driver/docker.go]: Does ```strings.Split(driverConfig.ImageName, "/")[0]``` return the repo? In that case we could just use ```repo```

**Comment by diptanu** [comment]
@ErikEvenson Thanks for taking a stab at this! I left a few comments, I hope after a few iterations we can get it merged this and get it out for 0.3.1 which is going to come out in a few days.

File: client/driver/docker.go (hunk 3 above)
Comment (ErikEvenson) [inline client/driver/docker.go]: Will add this.

File: client/driver/docker.go (hunk 2 above)
Comment (ErikEvenson) [inline client/driver/docker.go]: This transformation has to occur before the image is pulled and inspected (around line 496) and therefore before the `createContainer` method is called (around line 543).

File: client/driver/docker.go (hunk 1 above)
Comment (ErikEvenson) [inline client/driver/docker.go]: Agreed -- will change.

File: client/driver/docker.go (hunk 4 above)
Comment (ErikEvenson) [inline client/driver/docker.go]: No, the split command returns the image string with the url scheme and without the url path -- which is different from the repo string.

**Comment by ErikEvenson** [comment]
I incorporated/answered your comments @diptanu and brought the PR up to date with master.  Cheers.

File: client/driver/docker.go (hunk 2 above)
Comment (diptanu) [inline client/driver/docker.go]: @ErikEvenson Oh I see. Can we create another method called ```Init()``` then? This could be called before Validate()

**Comment by diptanu** [comment]
@ErikEvenson I just had one last comment about introducing an ```Init``` method to do the transformation of the image name before Validate is being called. Once this is done, I think we can merge this PR.

**Comment by ErikEvenson** [comment]
@diptanu great -- I'll try to do this later today.

**Comment by ErikEvenson** [comment]
I have brought this PR up to date with master and made the changes @diptanu requested regarding the Init() function.  Note: master is not passing the `make test` test suite on my local box.

**Comment by steve-jansen** [comment]
@ErikEvenson out of curiosity, how does your Nomad cluster authenticate with ECR?  

My understanding is ECR requires a call to IAM for [GetAuthorizationToken](http://docs.aws.amazon.com/AmazonECR/latest/APIReference/API_GetAuthorizationToken.html), then invoke a docker login.  However, this token is only valid for 12 hours.

Are you handling this token out of band on your cluster?

**Comment by ErikEvenson** [comment]
@steve-jansen typically, before I run a nomad job, I have the cluster authenticate itself via a consul command.  So something like this:

`consul exec --http-addr=<IP_ADDR>:8500 "su docker -c '$(aws ecr get-login)'"`

**Comment by diptanu** [comment]
@ErikEvenson Oh I see. Can we improve this? I would like this to be a no-op and Nomad handling this for the user.

**Comment by diptanu** [comment]
@pshima Any thoughts here?

**Comment by ErikEvenson** [comment]
If you are running nomad on an EC2 host, you could give the EC2 instance a role that gives the instance full access to your ECR repos.  Then, you wouldn't have to store or even log in to AWS ECR.

**Comment by steve-jansen** [comment]
@ErikEvenson 

> Then, you wouldn't have to store or even log in to AWS ECR.

Very interesting.  Have you been able to bypass the need for `aws ecr get-login` (or equivalent API call) with an IAM EC2 instance profile?

**Comment by ErikEvenson** [comment]
@steve-jansen I haven't actually used this technique with nomad as it is a sort of lock-in to the AWS platform.  I would like my nomad scripts to be just as useful on other cloud provider platforms.  My expectation is that you would not need to do `aws ecr get-login` prior to pulling from the repo though.

**Comment by diptanu** [comment]
@ErikEvenson This is something we might be able to pull off by having the concept of image providers for different type of environments - so we could have a image provider which works well with AWS ECR and another with GCP for example.

**Comment by ErikEvenson** [comment]
@diptanu Right, but you are always going to want to have the ability to, say, pull a docker image from AWS for use on a, say, Azure instance.

**Comment by jaygorrell** [comment]
I didn't test this but I don't believe it's accurate to say you don't need to use `aws ecr get-login` if you have an instance profile. That would only mean that you don't need a credentials file for that command to work.

The point of the command is still to run a `docker login` command.

**Pull request merged**